gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java

```diff
@@ -124,6 +124,32 @@ public synchronized int read(@Nonnull byte[] buf, int offset, int length) throws
     return response;
   }
 
+  @Override
+  public void readFully(long position, byte[] buffer, int offset, int length) throws IOException {
+    streamStatistics.readFullyOperationStarted(position, length);
+    checkNotNull(buffer, "buf must not be null");
+    validatePositionedReadArgs(position, buffer, offset, length);
+    if (length == 0) {
+      return;
+    }
+    int nread = 0;
+    synchronized (this) {
+      long oldPos = getPos();
+      try {
+        seek(position);
+        while (nread < length) {
+          int nbytes = read(buffer, offset + nread, length - nread);
+          if (nbytes < 0) {
```

Comment:
if it is 0, it is not EOF?

Reply:
No.

Reply:
Can this cause infinite loop?

Reply:
Yes - if your server is not returning data, anyways it will be infinite loop for application. By default today also readFully can cause infinite waiting if server keeps on returning 0 bytes.